lib/src/vs/oracle.rs

```diff
@@ -0,0 +1,167 @@
+	SysTimeCounter(SysTimeCounter),
+	// EpochCounter versionstamp oracle provides a vector clock that uses the "epoch" as the first logical time
+	// and the in-memory counter for the second logical time.
+	//
+	// The epoch is supposed to be persisted in the underlying KVS and increased by one on each database restart.
+	// The in-memory counter resets on each database restart and increased by one on each now() call.
+	//
+	// EpochCounter is designed to be used instead of the SysTimeCounter when the runtime environment
+	// does not provide a monotonic system clock, and the database is running in a single-node mode.
+	EpochCounter(EpochCounter),
+}
+
+impl Oracle {
+	pub fn now(&mut self) -> Versionstamp {
+		match self {
+			Oracle::SysTimeCounter(sys) => sys.now(),
+			Oracle::EpochCounter(epoch) => epoch.now(),
+		}
+	}
+}
+
+pub struct SysTimeCounter {
+	// The first element is the saved physical time of the last versionstamp.
+	// The second element is the in-memory counter that resets every second.
+	state: Mutex<(u64, u16)>,
+
+	stale: (u64, u16),
+}
+
+impl SysTimeCounter {
+	pub fn now(&mut self) -> Versionstamp {
+		// Increment the counter and get the current number as the logical time
+		// only if the current physical time is the same as the last physical time.
+		// Otherwise, reset the counter to 0 and get the current number as the logical time.
+		// This is to ensure that the logical time is always increasing.
+		let state = self.state.lock();
+		if let Ok(mut state) = state {
+			let (last_physical_time, counter) = *state;
+			let current_physical_time = secs_since_unix_epoch();
+			let current_logical_time = if last_physical_time == current_physical_time {
+				counter
+			} else {
+				state.1 = 0;
+				0
+			};
+			state.0 = current_physical_time;
+			state.1 += 1;
+			self.stale = (current_physical_time, current_logical_time);
+			u64_u16_to_versionstamp(current_physical_time, current_logical_time)
+		} else {
+			u64_u16_to_versionstamp(self.stale.0, self.stale.1)
+		}
+	}
+}
+
+// EpochCounter versionstamp oracle providers a vector clock uses the "epoch" as the first time
+// and the in-memory counter for the second logical time.
+// The epoch is supposed to be persisted in the underlying KVS and increased by one on each database restart.
+// The in-memory counter resets on each database restart and increased by one on each now() call.
+// TODO: Refer to a paper that describes this concept and use the correct terminology.
+pub struct EpochCounter {
+	// epoch is the first half of the versionstamp that persists
+	// until the database restarts.
+	// This needs to be persisted externally so that the epoch increases by one
+	// on each database restart.
+	epoch: u16,
+
+	// state is the in-memory counter increases by one on each now() call.
+	counter: AtomicU64,
+}
+
+impl EpochCounter {
+	pub fn now(&mut self) -> Versionstamp {
+		let counter = self.counter.fetch_add(1, Ordering::SeqCst);
```

Comment:
Rust noob here, I saw that fetch_add (Adds to the current value, returning the previous value). This means it is returning the previous value of counter, and not the current value. Am i right in assuming this? Could lead to stale time?

Reply:
@arriqaaq Thanks! You are absolutely correct on how fetch_add works.

I think this implementation is correct though, because `now()` is the only way to grab a new versionstamp. By using fetch_add this way, we basically treat the "previous value" as the current versionstamp and return it, preparing the next versionstamp at the same time.

Reply:
Oh I see, thank you for the explanation!

Reply:
Can we just add a unit test case just to verify this?

Reply:
Hey @arriqaaq, Thanks a lot for your confirmation! Re: test, does this one seem reasonable to you?
https://github.com/surrealdb/surrealdb/blob/b775bdab05d3a0f80df954a103a321dd4df06b7d/lib/src/vs/oracle.rs#L153-L169

Reply:
Thank you @mumoshu, this does look reasonable, appreciate it much.

Reply:
Thanks a lot for your confirmation @arriqaaq!